cloudinit/config/cc_grub_dpkg.py

```diff
@@ -62,22 +67,35 @@ def handle(name, cfg, _cloud, log, _args):
     idevs_empty = util.get_cfg_option_str(
         mycfg, "grub-pc/install_devices_empty", None)
 
-    if ((os.path.exists("/dev/sda1") and not os.path.exists("/dev/sda")) or
-       (os.path.exists("/dev/xvda1") and not os.path.exists("/dev/xvda"))):
+    # get the root disk where the /boot directory resides.
+    # the ESP will reside on the same root disk as /boot, but not device.
+    try:
+        disk = util.subp(['grub-probe', '-t', 'disk', '/boot'],
+                         capture=True)[0].strip()
+        # if /dev/disk/by-id/ exists, attempt to determine robust name
```

Comment:
If you have the disk path, instead of guessing at what symlnks might be present, you can query udev for devlinks that are present:

```
% udevadm info --query=symlink /dev/nvme0n1
disk/by-id/nvme-SAMSUNG_MZVLB1T0HALR-000L7_S3TPNY0JB00151 disk/by-path/pci-0000:05:00.0-nvme-1 disk/by-uuid/3792634682180574674 disk/by-id/nvme-eui.0025388b710116a1 disk/by-label/rpool
```

Reply:
This is absolutely the way to go. Thanks! Implemented in latest revision.